src/braket/default_simulator/linalg_utils.py

```diff
@@ -50,15 +45,22 @@ def multiply_matrix(
         return _multiply_matrix(state, matrix, targets)
 
     control_state = control_state or (1,) * len(controls)
-    num_qubits = len(state.shape)
-    control_slices = {i: _SLICES[state] for i, state in zip(controls, control_state)}
-    ctrl_index = tuple(
-        control_slices[i] if i in controls else _NO_CONTROL_SLICE for i in range(num_qubits)
-    )
-    state[ctrl_index] = _multiply_matrix(state[ctrl_index], matrix, targets)
+
+    ctrl_slices = [slice(None)] * len(state.shape)
+    for i, state_val in zip(controls, control_state):
+        ctrl_slices[i] = slice(None, 1) if state_val == 0 else slice(1, None)
+
+    state[tuple(ctrl_slices)] = _multiply_matrix(state[tuple(ctrl_slices)], matrix, targets)
     return state
 
 
+@cache
+def _compute_inverse_permutation(targets: tuple, num_qubits: int):
```

Comment:
Out of curiosity, how much does this save us? This doesn't seem to be that that expensive, even if called for each operation.

Reply:
Testing it out, saved around 1-4% depending on the circuit size.